core/src/main/clojure/xtdb/types.clj

```diff
@@ -1130,21 +1337,48 @@
     (cond
       (= 1 (count col-types)) (first col-types)
       (set/subset? col-types #{:float4 :float8}) :float8
-      (set/subset? col-types #{:int2 :int4 :int8}) :int8)))
+      (set/subset? col-types #{:int2 :int4 :int8}) :int8
+      (is-list-type col-types) [:list (->unified-col-type (second col-types))])))
+
+(defn remove-nulls
+  [typ]
+  (if (is-list-type typ)
+    (let [inner (-> typ
+                    second
+                    (disj :null))]
+      ;; wrap in extra vector as next step maps - we want the whole [:list :inner] evaluated
+      [[:list (if (= 1 (count inner))
+                (first inner)
+                inner)]])
+    (disj typ :null)))
 
 (defn field->pg-type
   ([field] (field->pg-type nil field))
 
   ([fallback-pg-type ^Field field]
    (let [field-name (.getName field)
          col-type (field->col-type field)
-         col-types (-> (flatten-union-types col-type)
-                       (disj :null)
+         col-types (-> ((if (is-list-type col-type)
+                          flatten-list-unions
+                          flatten-union-types) col-type)
+                       remove-nulls
                        (->> (into #{} (map (partial col-type->pg-type fallback-pg-type)))))]
+     (log/tracef "field->pg-type %s, col-type %s, field-name %s" (pr-str col-types) col-type field-name)
      (-> (if-let [col-type (->unified-col-type col-types)]
            col-type
            (col-type->pg-type fallback-pg-type col-type))
          (pg-types)
          (set/rename-keys {:oid :column-oid})
          (assoc :field-name field-name
                 :column-name field-name)))))
+
```

Comment:
we don't currently tend to check these in - they tend to become tests in the test suite (if appropriate)